### ISSUE TITLE
JSON schema validation passes when it shouldn't

### ISSUE THREAD

**Issue by DannyBen** [DannyBen/bashly#465]
It appears that our `bashly.json` schema allows invalid value types to pass as valid.

The below is how it *should* behave:

### Schema

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "default": {
      "type": "string"
    }
  },
  "required": ["default"],
  "additionalProperties": false
}
```

### Test file

```yaml
# default: some string
default:
  - one
  - two
```

### Result

**Fails as expected** when running this command:

```
$ check-jsonschema --schemafile schema.json test.yml
Schema validation errors were encountered.
  test.yml::$.default: ['one', 'two'] is not of type 'string'
```

*(this test was done to rule out check-jsonschema as the culprit)*

---

However, when using disallowed value types in `bashly.yml`, the validation passes:

### bashly.yml

```yaml
# bashly.yml
default:
  key1: value
  key2: value
```

### Result

```
$ check-jsonschema --schemafile bashly.json bashly.yml
ok -- validation done
```

In fact, anything I use seem to pass validation, even this `bashly.yml`

```yaml
asd: asd
foo: bar
```

**Comment by DannyBen** [comment]
The problem lies within the `custom-properties` definition.

This schema will pass validation on any given file.

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "custom-properties": {
      "patternProperties": {
        "^x_.": {
          "title": "custom property"
        }
      }
    }
  },
  "type": "object",
  "properties": {
    "default": {
      "type": "string"
    }
  },
  "required": ["default"],
  "additionalProperties": false,
  "$ref": "#/definitions/custom-properties"
}
```

when tested with check-jsonschema. I wonder if it is the same when tested with other validators.

**Comment by DannyBen** [comment]
Quoting from [this page](https://json-schema.org/understanding-json-schema/structuring#dollarref):

> In Draft 4-7, $ref behaves a little differently. When an object contains a $ref property, the object is considered a reference, not a schema. Therefore, **any other properties you put in that object will not be treated as JSON Schema** keywords and will be ignored by the validator. $ref can only be used where a schema is expected.

If I am reading this right, it means we cannot use `$ref` to include a partial property, no?

**Comment by DannyBen** [comment]
Wow, this JSON schema thing is so ugly......

I believe we should be using `allOf`:

```json
      "items": {
        "allOf": [
          { "$ref": "#/$defs/argument" },
          { "$ref": "#/$defs/custom-properties" }
        ]
      }
```

as hinted by [this SO page](https://stackoverflow.com/questions/29352100/reusing-an-object-for-multiple-json-schemas), but I was unable to make it work both for inner elements (argument) and root.

If needed, I can make a much simpler YAML template that can be converted to JSON without using `$ref` at all - but I am done for today, will wait for your input.

**Comment by DannyBen** [comment]
One thing that I found out that is working, is to remove the definition of the `custom-properties` and instead, use the `patternProperties` snippet as is multiple times. Then, the schema seems to be working as intended.

Should I do it?

**Comment by EmilyGraceSeville7cf** [comment]
> Wow, this JSON schema thing is so ugly......
> 
> I believe we should be using `allOf`:
> 
> ```json
>       "items": {
>         "allOf": [
>           { "$ref": "#/$defs/argument" },
>           { "$ref": "#/$defs/custom-properties" }
>         ]
>       }
> ```
> 
> as hinted by [this SO page](https://stackoverflow.com/questions/29352100/reusing-an-object-for-multiple-json-schemas), but I was unable to make it work both for inner elements (argument) and root.
> 
> If needed, I can make a much simpler YAML template that can be converted to JSON without using `$ref` at all - but I am done for today, will wait for your input.

I was thinking about creating something like a JSON schema generator from YAML. I guess it's a good take, but it should be a separate project that can be incorporated in this one.

**Comment by EmilyGraceSeville7cf** [comment]
> One thing that I found out that is working, is to remove the definition of the `custom-properties` and instead, use the `patternProperties` snippet as is multiple times. Then, the schema seems to be working as intended.
> 
> Should I do it?

Or we can use `allOf`. The problem is that VS Code tells nothing about this issue for instance and the schema seems to work inside it...

**Comment by DannyBen** [comment]
> I was thinking about creating something like a JSON schema generator from YAML.

This is easily done. I have already tried it with `yq` for converting the JSON schema to YAML, replacing the `$ref` with YAML anchor, and then regenerating it back. Should I post it in a branch for review?

> Or we can use `allOf`.

I was unable to make it work for the root element.

> The problem is that VS Code tells nothing about this issue

The quote I pasted about `$ref` overriding everything else seems to be the definitive guide as I understand.

**Comment by EmilyGraceSeville7cf** [comment]
> YAML Language support uses [JSON Schemas](https://json-schema.org/) to understand the shape of a YAML file, including its value sets, defaults and descriptions. The schema support is shipped with JSON Schema Draft 7.

Quote from [this](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml) page. But the following YAML for some reason is still not valid:

```yaml
aa: 1
default: a
```

for:

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "custom-properties": {
      "patternProperties": {
        "^x_.": {
          "title": "custom property"
        }
      }
    }
  },
  "type": "object",
  "properties": {
    "default": {
      "type": "string"
    }
  },
  "required": ["default"],
  "additionalProperties": false,
  "$ref": "#/definitions/custom-properties"
}
```

**Comment by EmilyGraceSeville7cf** [comment]
Or we can try to use [this](https://github.com/santhosh-tekuri/jsonschema) validator to match VS Code editor behavior as much as possible.

**Comment by DannyBen** [comment]
👉 #468

**Comment by DannyBen** [comment]
> we can try to use [this](https://github.com/santhosh-tekuri/jsonschema) validator to match VS Code editor behavior

Matching to VS Code validation is in no way a priority for me. Our schema should simply be standards compliant. If VS code validation works as intended (i.e., it FAILS when there are arbitrary values that do not start with `x_`), then it means that either CS code is wrong, or `check-jsonschema` is wrong.

I tested `jv` and judging just by the 10 minutes of test, it is not better:

1. It also does not complain when testing the below YAML against our currently live bashly schema
2. It does not allow reading YAML with anchors (like I have with almost all of my bashly.yml files).

**Comment by DannyBen** [comment]
Fixed in #468 + #467
